rust_icu_ubrk/src/lib.rs

```diff
@@ -0,0 +1,735 @@
+//! are available.
+//!
+//! ## Examples
+//!
+//! Sample code use is given below.
+//! 
+//! ```rust
+//! use rust_icu_sys as sys;
+//! use rust_icu_ubrk as ubrk;
+//!
+//! let text = "The lazy dog jumped over the fox.";
+//! let mut iter = ubrk::UBreakIterator::try_new(
+//!     sys::UBreakIteratorType::UBRK_WORD, "en", text).unwrap();
+//!
+//! assert_eq!(iter.last_boundary(), 33);
+//! assert_eq!(iter.current(), 33);
+//! assert!(iter.is_boundary(13));
+//! assert!(iter.is_boundary(19));
+//! assert!(!iter.is_boundary(15));
+//! assert_eq!(iter.following(15), 19);
+//! assert_eq!(iter.current(), 19);
+//! assert_eq!(iter.preceding(15), 13);
+//! assert_eq!(iter.current(), 13);
+//! assert_eq!(iter.previous(), Some(12));
+//! assert_eq!(iter.current(), 12);
+//! 
+//! // Reset to first boundary and consume `iter`.
+//! assert_eq!(iter.first(), 0);
+//! let breaks: Vec<i32> = iter.collect();
+//! assert_eq!(breaks, vec![3, 4, 8, 9, 12, 13, 19, 20, 24, 25, 28, 29, 32, 33]);
+//! ```
+//!
+//! See the [ICU user guide](https://unicode-org.github.io/icu/userguide/boundaryanalysis/)
+//! and the C API documentation in the
+//! [`ubrk.h` header](https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/ubrk_8h.html)
+//! for details.
+
+use {
+    rust_icu_common::{self as common, simple_drop_impl},
+    rust_icu_sys::{self as sys, *},
+    rust_icu_ustring as ustring,
+    std::{convert::TryFrom, ffi, os::raw, ptr, rc::Rc},
+};
+
+/// Returned by break iterator to indicate that all text boundaries have been returned.
+// UBRK_DONE is defined as a macro in ICU and macros are not currently supported
+// by bindgen, so we define it ourselves here.
+pub const UBRK_DONE: i32 = -1;
+
+/// Rust wrapper for the ICU `UBreakIterator` type.
+pub struct UBreakIterator {
+    // Pointer to the underlying ICU representation.
+    rep: ptr::NonNull<sys::UBreakIterator>,
+
+    // The underlying C representation holds pointers to `text` and exactly one of
+    // {`locale`, `rules`, `binary_rules`} throughout its lifetime. We are
+    // responsible for ensuring that the pointers remain valid during that time,
+    // and for dropping the referenced values once the underlying C representation
+    // is released.
+    //
+    // A break iterator may be cloned, in which case the underlying C representation
+    // of the cloned iterator will hold pointers to the values pointed to by the
+    // original iterator, while maintaining its own internal iteration state.
+    //
+    // For these reasons, these fields are wrapped in `Rc`, ensuring that the
+    // referenced values (`text`, `locale`, etc.) are not released prematurely
+    // if the original break iterator is dropped before its clone. As break
+    // iterators are inherently not thread-safe [1], `Rc` was chosen over `Arc`.
+    //
+    // [1] https://unicode-org.github.io/icu/userguide/boundaryanalysis/#thread-safety
+    text: Rc<ustring::UChar>,
+    locale: Option<Rc<ffi::CString>>,
```

Comment:
Note, you could use `ULoc` here instead of `ffi::CString` in case that helps handling the underlying C string.

Reply:
Great. I hadn't thought of that. Addressed as part of 70f4642.

Reply:
Great. I hadn't thought of that. Addressed as part of 70f4642.

Reply:
Correction: I added an iterator that exposes available locales as `ULoc` in 70f4642.

I have just now updated `try_new_ustring` to accept locale as a `ULoc` (2fce06f), which I think makes more for a better API, but ultimately the value that is owned by the iterator is the `CString`, so I did not change the type of the field in the struct.

Reply:
OK.